datafusion/expr/src/type_coercion/binary_result_type.rs

```diff
@@ -0,0 +1,150 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Coercion rules for matching argument types for binary operators
+
+use arrow::datatypes::DataType;
+
+use datafusion_common::DataFusionError;
+use datafusion_common::Result;
+
+use crate::type_coercion::binary::{
+    any_decimal, bitwise_coercion, decimal_op_mathematics_type, math_decimal_coercion,
+    mathematics_numerical_coercion, mathematics_temporal_common_type,
+    string_concat_coercion,
+};
+use crate::type_coercion::{is_datetime, is_interval};
+use crate::Operator;
+
+/// returns the resulting type of a binary expression evaluating the `op` with the left and right hand types
```

Comment:
I think this code looks good but I am +0 on putting it in its own module -- since it needs to remain in sync with `binary.rs` it may make sense to keep the code there or maybe in `datafusion/expr/src/type_coercion/binary/result_type.rs`

This is a minor point and we can reorganize the code in some future PR